cpp/ql/lib/semmle/code/cpp/models/implementations/Win32CommandExecution.qll

```diff
@@ -0,0 +1,56 @@
+private import semmle.code.cpp.models.interfaces.CommandExecution
+
+/** The `ShellExecute` family of functions from Win32. */
+class ShellExecute extends Function {
+  ShellExecute() { this.hasGlobalName("ShellExecute" + ["", "A", "W"]) }
+}
+
+private class ShellExecuteModel extends ShellExecute, CommandExecutionFunction {
+  override predicate hasCommandArgument(FunctionInput input) { input.isParameterDeref(2) }
+}
```

Comment:
Is there a specific reason to split this up into two classes?

Reply:
A while back (many years ago, actually) we decided that we'd like to have the charpred exposed as a public class so that users could work with the functions without having to name them manually. i.e., they should be able to do:
```ql
from ShellExecute shellExec
where ...
select shellExec.getACall(), "This is a bad function to call!"
```
However, we didn't want to expose the fact that `ShellExecute` extended various model-related classes.

So the solution we came up with back then was this split between the public class (which implemented no models), and a private class which implemented all the required models

Reply:
Thanks for the clarification on the internal issue.